### DISCUSSION TITLE
Make `set_script(script)` load `script` as a path if `script` is a String

### DISCUSSION THREAD

**Discussion by InkONat** [godotengine/godot-proposals#8064] [Engine Core]
I recently had to bug fix something in my project and had to get help from the Godot's discord server's #advanced channel. The problem was that `set_script(script)` can take any variable type for `script`, and therefore it could take a String for `script`, so I thought that it internally loaded the script at the path `script` if `script` was a String

However, that was not the case, and me and the helper wasted some time trying to find the issue, so I suggest that Object's `set_script(script)` loads a script at path `script` if `script`'s type is a String, for example:

`set_script("path/to/script.gd")` would now work and set the Object's script to script.gd, as the function now would load that string and then set the script to what was loaded

**Comment by InkONat** [comment]
Sorry for any grammar mistakes or typos, I'm not a native english speaker

**Comment by Mickeon** [comment]
Unless the path is always absolute, this cannot be done. Besides, `set_script(preload("path/to/script.gd"))` is more explicit and consistent. I'm not sure how to explain it better, but I'll give it a try.
**Script**s are Resources like many. Even if the engine treats them a bit more specially, they can be saved, loaded and stored as properties like all others. Therefore, using `load`/`preload` to fetch them from disk makes sense because all Resources work that way, too.

>  set_script(script) can take any variable type for script, and therefore it could take a String for script

Hey..... https://github.com/godotengine/godot/issues/81734

**Comment by AThousandShips** [comment]
This would create a bit of a confusion as it would be the only place you could do this, and as pointed out in godotengine/godot#81734 this is due to a declaration order thing where `Script` isn't available, and while it can be forward declared here it wasn't at the time this code was added, so the fact that it is `Variant` isn't because that is desired, but because it solves an issue

**Reply by Mickeon** [reply]
To play devil's advocate, the already existing and only error message is a bit problematic. Scripts are not Strings, but you can pass absolutely anything in the method, without failing at compile-time. When `set_script` inevitably fails, what the error says is not exactly clear, especially if one has just started programming. The user is left in their own thoughts, so I can understand the confusion.

The least that could be done is an error message for a **String**/**StringName** type, (not present in any release build, of course).

**Reply by AThousandShips** [reply]
The confusion I mean would be people assuming they can use `set_texture("res://my_texture.png")` elsewhere, beyond just the complexity of accomplishing this

Again this is a complex case and the solution would be to eventually change the argument type, not to expand the functionality in an assymetric way

**Reply by Mickeon** [reply]
I do not see the harm in adding a more concrete error message(s) for the time being. The issue is so nested to the core of the engine, and it's such a low priority in the grand scheme of things, that it may take a while to actually fix it.
